### PR TITLE
adding notebook for text generation using transformers

### PR DESCRIPTION
**Pull request by sanjanalreddy**
This notebook requires a Vertex AI notebook instance with TF2.12 and a GPU.

You will also need to install `keras-nlp`. The notebook has instructions for this. 

Note: I haven't added `keras-nlp` to requirements.txt because I want to keep the setup simple for Next'23. Once next is completed, we can add this to requirements.txt.

### CODE REVIEW

File: notebooks/text_models/solutions/text_generation_using_transformers.ipynb [hunk 1]
```diff
@@ -0,0 +1,911 @@
+{
```
Comment (BenoitDherin) [inline notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <ul><li>scaled down Generative --&gt; scaled down generative</li><li>Maybe mention the number of books contained in simplebooks-92 to give an idea of the size of the dataset</li><li>This demonstrate --&gt; This notebook demonstrate</li><li>Add a "Learning objective" section to the overview with bullet points to match the style of the other notebooks (it seems the last sentence of this paragraph lists the learning objectives)</li></ul><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/372/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (BenoitDherin) [inline notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <ul><li>Something off in this sentence: "Keras NLP offers transformer later which..." later?</li></ul><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/372/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (BenoitDherin) [inline notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Maybe give a very brief explanation about what these hyper-parameters are about. In particular, what is the difference between <code>SEQ_LEN</code> and <code>MIN_TRAINING_SEQ_LEN</code>?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/372/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (BenoitDherin) [inline notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>In <code>dir = os.path.expanduser</code> maybe use a different variable name (e.g. <code>data_path</code> )  than <code>dir</code>, since <code>dir</code> is a reserved keyword.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/372/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (BenoitDherin) [inline notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <ul><li>We want to limit the vocabulary as much as possible, as we will see later on that it has a large effect on the number of model parameters. --&gt; We want to limit the vocabulary as much as possible, since it has a large effect on the number of model parameters.</li><li><em>too few</em> vocabulary terms --&gt; <em>too few</em> words</li><li>maybe add a quick sentence to describe what the <code>WordPieceTokenizer</code> is and does (a simple example of a "WordPieceTokenized" word would be great)</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/372/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (BenoitDherin) [inline notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Maybe add a brief sentence before this cell to explain that 1) the goal will be to predict a "label sequence" of "next words" from a "features sequence" of "previous words", and 2) the "feature" is obtained by shifting the original sequences to the right using the [BOS] token.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/372/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (BenoitDherin) [inline notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <ul><li>"scaled down transformer decoder based generative text model" --&gt; "scaled-down transformer-decoder-based generative text model" (maybe too many qualifying words before "model" here)</li><li>Explain what "causal masking" is</li><li>I'd remove the sentence "The layer has no cross-attention when run with decoder sequence only." which reads a bit confusing without more explanation.</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/372/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (BenoitDherin) [inline notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Super cool!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/372/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (BenoitDherin) [inline notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <ul><li>"To start lets" --&gt; "To start, let us"</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/372/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (BenoitDherin) [inline notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <ul><li>Notes: --&gt; <strong>Notes:</strong></li></ul><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/372/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (BenoitDherin) [inline notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Maybe add a brief comment on why we would want to do that (e.g. debugging? to see that the model improves after each epoch?)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/372/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='42'/>

Comment (BenoitDherin) [inline notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Great!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/372/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='44'/>

**Review by BenoitDherin — APPROVED** [review]
LGTM. Really great new addition! Thanks Sanjana!

File: notebooks/text_models/solutions/text_generation_using_transformers.ipynb (hunk 1 above)
Comment (takumiohym) [reply notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>+1!</p>

Comment (takumiohym) [inline notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Could you elaborate a bit more about Decoder-Only model?</p><p>I think it's good to mention how it is different from the encoder-decoder approach in terms of text generation, and touch some actual model names (e.g. T5 vs GPTs)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/372/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (takumiohym) [inline notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Could you add hyperlinks to the keras document for each keras_nlp module in markdown? </p><p>https://keras.io/api/keras_nlp/tokenizers/word_piece_tokenizer/</p><p> </p><p>Same for other modules.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/372/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (takumiohym) [inline notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Perhaps worth mentioning that this <code>TokenAndPositionEmbedding</code> creates trainable positional embedding instead of the fixed sinusoidal embedding which original Transformer was using.</p><p> </p><p>Also, I'd add what components are contained in <code>TransformerDecoder</code> with a diagram from the paper if appropriate. </p><p>Adding the link to the source code (or Keras document) may be helpful for advanced and curious students.</p><p>https://github.com/keras-team/keras-nlp/blob/v0.6.1/keras_nlp/layers/modeling/transformer_decoder.py#L31</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/372/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (takumiohym) [inline notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Great explanation! I love this inference section :)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/372/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='40'/>

**Review by takumiohym — APPROVED** [review]
Added a few minor comments, but everything is related to the markdown descriptions.
Great addition! Thanks Sanjana!

File: notebooks/text_models/solutions/text_generation_using_transformers.ipynb (hunk 1 above)
Comment (sanjanalreddy) [reply notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Done</p>

Comment (sanjanalreddy) [reply notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Good catch! Updated.</p>

Comment (sanjanalreddy) [reply notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Thank you! Updated! </p>

Comment (sanjanalreddy) [reply notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Thanks! </p>

Comment (sanjanalreddy) [reply notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Done.</p>

Comment (sanjanalreddy) [reply notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Done</p>

Comment (sanjanalreddy) [reply notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Thanks!</p>

Comment (sanjanalreddy) [reply notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Absolutely! Added an explanation along with link to their origin papers.</p>

Comment (sanjanalreddy) [reply notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Done!</p>

Comment (sanjanalreddy) [reply notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Done!</p>

Comment (sanjanalreddy) [reply notebooks/text_models/solutions/text_generation_using_transformers.ipynb:1]: <p>Nice suggestion. I have added the hyperlinks and updated the description of the embedding layer. Unfortunately the original paper doesn't have a decoder only model represented and showing both the encoder and the decoder will be confusing so for now, I'll leave out the diagram.</p>
